### PR TITLE
Native support for Amazon Lambda

### PR DESCRIPTION
**Pull request by patriot1burke**
* Still have to generate yaml files from OutputTargetBuildItem
* Needs some cleanup
* Needs doc update too

### CODE REVIEW

File: extensions/amazon-lambda/deployment/src/main/java/io/quarkus/amazon/lambda/deployment/LambdaConfig.java [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
+package io.quarkus.amazon.lambda.deployment;
 
 import java.util.Optional;
 
 import io.quarkus.runtime.annotations.ConfigItem;
 import io.quarkus.runtime.annotations.ConfigPhase;
 import io.quarkus.runtime.annotations.ConfigRoot;
 
-@ConfigRoot(phase = ConfigPhase.RUN_TIME)
+@ConfigRoot(phase = ConfigPhase.BUILD_TIME)
```
Comment (stuartwdouglas) [inline extensions/amazon-lambda/deployment/src/main/java/io/quarkus/amazon/lambda/deployment/LambdaConfig.java]: This should really be runtime. The idea with the runtime config of the handler name was that you could create a single native image with multiple handlers, and use environment variables to select the handler to use at deployment time.

If this is build time and you have a deployment that has 10 different functions you would need 10 different native image builds.

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]

File: extensions/amazon-lambda/deployment/src/main/java/io/quarkus/amazon/lambda/deployment/LambdaConfig.java (hunk 1 above)
Comment (patriot1burke) [reply extensions/amazon-lambda/deployment/src/main/java/io/quarkus/amazon/lambda/deployment/LambdaConfig.java]: You shouldn't have a deployment with 10 different functions.

Comment (patriot1burke) [reply extensions/amazon-lambda/deployment/src/main/java/io/quarkus/amazon/lambda/deployment/LambdaConfig.java]: Moving this back to runtime would also require moving error condition checks from build time to runtime which can be very problematic.  You don't want to have to debug build time issues at runtime.

File: docs/src/main/asciidoc/amazon-lambda-http-guide.adoc [hunk 2]
```diff
@@ -3,36 +3,230 @@ This guide is maintained in the main Quarkus repository
...
+for installing AWS SAM CLI.
+
+== Creating the Maven Deployment Project
+
+Create the Quarkus AWS Lambda maven project using our Maven Archetype.
+
+
+[source, subs=attributes+]
+----
+mvn archetype:generate \
```
Comment (stuartwdouglas) [inline docs/src/main/asciidoc/amazon-lambda-http-guide.adoc]: This is fine for now as we don't have anything better, but eventually I hope we can have better create-project tooling that can be extended to handle cases like this.

File: extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/LambdaConfig.java [hunk 3]
```diff
@@ -19,6 +19,14 @@
...
-    Optional<String> handler;
+    public Optional<String> handler;
+
+    /**
+     * If true, this will enable the aws event poll loop within a Quarkus test run. This loop normally only runs in native
+     * image. This option is strictly for testing purposes.
+     *
+     */
+    @ConfigItem
+    public boolean enablePollingJvmMode;
```
Comment (stuartwdouglas) [inline extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/LambdaConfig.java]: Do we need this to be config, can't we handle this transparently? e.g. have the io.quarkus.amazon.lambda.test.LambdaResourceManager set a system property, and just look at the system property to decide if we are in a test environment.

Comment (stuartwdouglas) [reply extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/LambdaConfig.java]: Either way it is not worth holding up the PR over it.

**Review by stuartwdouglas — APPROVED** [review]

File: extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/LambdaConfig.java (hunk 3 above)
Comment (patriot1burke) [reply extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/LambdaConfig.java]: I'll try that out in another PR.

**Pull request merged**